### PR TITLE
Use scratch (slim) container

### PR DESCRIPTION
**Pull request by im-kulikov**

### CODE REVIEW

**Comment by shenli** [comment]
@im-kulikov Thanks for your PR!
Could you describe what's the benefit of using scratch?

@iamxy @tennix PTAL

**Comment by im-kulikov** [comment]
@shenli `scratch` is empty container without dependencies
we don't need golang or other libs after building container

**Comment by im-kulikov** [comment]
mb it would be useful for the community

**Comment by iamxy** [comment]
@im-kulikov Thank you for your PR.
Actually we have integrated a Dockerfile(alpine image based) within TiDB CI workflow, and the Docker images of TiDB/TiKV/PD are pushed to Dockerhub automatically every day after a series of stability tests.
You can pull images from Dockerhub directly.
```
docker pull pingcap/tidb:latest
docker pull pingcap/tikv:latest
docker pull pingcap/pd:latest

# for release-1.0.3
docker pull pingcap/tidb:v1.0.3
docker pull pingcap/tikv:v1.0.3
docker pull pingcap/pd:v1.0.3
```

**Comment by im-kulikov** [comment]
@iamxy, thanks, but this images has third-party dependence’s (alpine/golang).. 
if need, I can push PR’s to all repos

As I say before ‘more dependencies - more potential vulnerabilities’

**Comment by iamxy** [comment]
Alright, thanks for your suggestion at risk.
As golang program, tidb-server runs well in an empty scratch Docker environment.
But another problem is tikv-server which needs dependency of `glibc`.
Could you file a PR for TiKV repo as well,  thinks very much!

File: Dockerfile [hunk 1]
```diff
@@ -1,11 +1,20 @@
-FROM golang:1.9
+# Builder image
+FROM golang:1.9 as builder
```
Comment (iamxy) [inline Dockerfile]: Suggest to use `golang:1.9.2` instead.

File: Dockerfile [hunk 2]
```diff
@@ -1,11 +1,20 @@
...
 
-RUN cd /go/src/github.com/pingcap/tidb && \
-    make && \
-    mv bin/tidb-server /tidb-server && \
-    make clean
+WORKDIR /go/src/github.com/pingcap/tidb/
+
+RUN make
+
+RUN chmod +x /go/src/github.com/pingcap/tidb/bin/tidb-server
```
Comment (iamxy) [inline Dockerfile]: Is this necessary ?

File: Dockerfile [hunk 3]
```diff
@@ -1,11 +1,20 @@
...
+
+# Executable image
+FROM scratch
+
+COPY --from=builder /go/src/github.com/pingcap/tidb/bin/tidb-server /tidb-server
+
+WORKDIR /
 
 EXPOSE 4000
```
Comment (iamxy) [inline Dockerfile]: I think ENTRYPOINT should be specified in the executable image.

File: Dockerfile (hunk 1 above)
Comment (im-kulikov) [reply Dockerfile]: It’s same:
1.9.2, 1.9, 1, latest: 1.9.2-stretch
With 1.9 we take latest 1.9.c

File: Dockerfile (hunk 3 above)
Comment (im-kulikov) [reply Dockerfile]: It’s not changed https://github.com/im-kulikov/tidb/blob/cd08d16bc27611da560bcd2421b723b47f00e085/Dockerfile#L21

File: Dockerfile (hunk 2 above)
Comment (im-kulikov) [reply Dockerfile]: I’ll check

File: Dockerfile (hunk 3 above)
Comment (iamxy) [reply Dockerfile]: Sorry, I missed it .

**Comment by im-kulikov** [comment]
Don’t merge pls, I’ll check ‘chmod’

**Comment by iamxy** [comment]
Actually I have verified that, it works well.
Thanks again for your effort :)

**Comment by im-kulikov** [comment]
last changes..
- use golang-alpine for building container
- add make, git

- https://hub.docker.com/r/library/golang/tags/1.9-stretch/
vs 
- https://hub.docker.com/r/library/golang/tags/1.9-alpine/

**Comment by iamxy** [comment]
LGTM

**Comment by ngaut** [comment]
LGTM

**Comment by ngaut** [comment]
Thank you for your contribution, @im-kulikov

**Comment by im-kulikov** [comment]
and thank you too

**Comment by im-kulikov** [comment]
Hi guys, can you help with this?

2a6f7eb7-5703-448d-82f5-32b053a61d31

**Comment by gingerkidney** [comment]
Hi @im-kulikov, I'm Yanqing, sorry to make you feel confused. 

Usually, we will inform every new contributor after the PR is merged by sending an email. And a mug with TiDB logo will be sent to the contributor to show our gratitude. If we didn't receive reply by email or the email is invalid then we will try other ways like linkedIn, Twitter, that's why I left a message on your Twitter. 

I can understand your concern. Sending a mug from China to Russia sounds not reliable, but we value each contribution. Our mug has been to different countries already like Sweden, America, Japan and so on. Some of them posted on Twitter like this:

https://twitter.com/dvrkps/status/935487535089881089

We will keep your personal information secret, just for sending the gift. But you don't have to fill the form if that makes you uncomfortable.

You did a great job and thank you for the contribution again.

**Comment by im-kulikov** [comment]
Oh.. it’s sounds good.. thanks

**Pull request merged**